Templating: Add bult in __user {name, id, login, email} variable to templating system

aidanmountford opened the pull request:
**Feature Add:**   Updated templating code to support the $__user variable to expose the current logged on userid, username, email and login.

**Bug Fix:**  Fixed the $__org variable as it was returning the user id instead of the orgId.

Closes #4922

aidanmountford committed:
Updated templating code to support the $__user variable to expose the current
userid, username, email and login.

Fixed the $__org variable as it was returning the user id instead of the ordId.

Updated the documentation to match

aknuds1 commented:
Frontend tests fail.

```
Executing 'yarn run typecheck'
yarn run v1.22.0
$ tsc --noEmit
public/app/features/templating/variable_srv.ts:76:35 - error TS2339: Property 'name' does not exist on type 'User'.

76             name: contextSrv.user.name,
                                     ~~~~

public/app/features/templating/variable_srv.ts:78:36 - error TS2339: Property 'email' does not exist on type 'User'.

78             email: contextSrv.user.email,
                                      ~~~~~
```

aidanmountford commented:
Yea i noticed that

Will fix it and resubmit

A

On Tue, Apr 7, 2020 at 5:36 PM Arve Knudsen <notifications@github.com>
wrote:

> Frontend tests fail.
>
> Executing 'yarn run typecheck'
> yarn run v1.22.0
> $ tsc --noEmit
> public/app/features/templating/variable_srv.ts:76:35 - error TS2339: Property 'name' does not exist on type 'User'.
>
> 76             name: contextSrv.user.name,
>                                      ~~~~
>
> public/app/features/templating/variable_srv.ts:78:36 - error TS2339: Property 'email' does not exist on type 'User'.
>
> 78             email: contextSrv.user.email,
>                                       ~~~~~
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafana/grafana/pull/23378#issuecomment-610227590>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACR6BDSSCGQKIBY4YADPUADRLLJZFANCNFSM4MCSNW7Q>
> .
>


-- 

*Aidan Mountford*
Co-Founder & CEO
Fuuzio Group Pty Ltd
T1.509-514, 55 Plaza Parade
Maroochydore, QLD, 4557
Mobile: 0417-849-852

-- 

**This email message is for the sole use of the intended recipient(s) and 
may contain confidential and privileged information. Any unauthorized 
review, use, disclosure or distribution is prohibited. If you are not the 
intended recipient, please contact the sender by reply email and destroy 
all copies of the original message.**
**
**

**All electrical contracting 
activities are undertaken by Fuuzio Operations Pty Ltd.  ABN 38 627 150 547 
 (Registered Electrical Contractor **NSW:333365C,  VIC:28883, SA:294101).   
 **
**
**
**Fuuzio Operations Pty. Ltd. and Foreman-Sheean Electrical & 
Communications Pty. Ltd. are wholly owned subsidiaries of Fuuzio Group Pty. 
Ltd.**

mckn commented on docs/sources/reference/templating.md:
```diff
@@ -386,6 +386,15 @@ This variable is the UID of the current dashboard.
+This variable is the ID of the current user.
+`${__org.name}` is the name of the current user.
+`${__org.login}` is the login of the current user.
+`${__org.email}` is the email of the current user.
```
Should this be changed to `__user.*` instead of the `__org.*`?

torkelo replied:
no, think its good that the options here are listed

aidanmountford replied:
Yea - that copy/paste at 4am strikes again...

Have updated the process so that it no longer relies on the properties being pulled into contextSvr from the user record - and rather pulls the info directly, fixed documentation, confirmed typecheck passes and integrated change into second declaration location.

Thanks for your input

aidanmountford replied:
Torkelo - mckn is correct re the __user*  instead of __org* in the documentation it should be __user.*   - If you would like me to expand this change to make available all properties of the user object - am happy to do so.

mckn replied:
I think it looks really good @aidanmountford!  Maybe also add the `__org.id` in the section about the `organizations`?

*Optional:* if you want to make it even better you can convert the (completeDashboardTemplating-thunk)[https://github.com/grafana/grafana/blob/6a5794aa7a44422c7c1e64d337dad33fab946e52/public/app/features/variables/state/actions.ts#L87] to `async/await` and instead of using a promise chain you can `await` the result from the `backendSrv`.

We can then also change so the (dispatch)[https://github.com/grafana/grafana/blob/c7ffc1199cdb331c34cf8d95a60ed602b14e76b5/public/app/features/dashboard/state/initDashboard.ts#L190] is being awaited as with the `processVariables` on the live above.

mckn replied:
I noticed that the tests are failing, please give me a ping if you need help sorting those out.

aidanmountford replied:
mckn :  I have moved the backend server call to an await before I read this for several reasons, one of which it makes the spec tests sort themselves out, and two, because it will be easier for the next person to simply duplicate the database request(s) and add a few lines of code to expand this.

I will expand the org, dashbaord and user objects to include all useful fields while I'm here - any other objects you believe might be worth exposing  -  I will do them at the same time - after all - its easter - and i am not supposed to leave the house (apparently).

I prb wont do much on this for 24 hours (only as I need to solve my immediate issue) and will focus on my real job (which has nothing to do with programming) for a day or two ..   Will bake this for you properly no later than Monday night.


I didn't have my environment setup to run the full tests - this was originally intended to be a quick change :)   But all good I fixed that last night 

I will sort out the dispatch

mckn replied:
wow, @aidanmountford this is awesome! Thanks for helping out in the project and please let me know if there is something I can do to assist you.

I will let you know if I can think of anything else that could be added here.

mckn reviewed (changes requested):
Love that you are helping out in the project @aidanmountford. Great work!

Overall I think i looks good but we are currently moving the template variables to a new redux structure and therefore we have two places that initialize the global variables. The old one will be removed soon but to have this properly working we need to add this to the following place as well:

https://github.com/grafana/grafana/blob/6a5794aa7a44422c7c1e64d337dad33fab946e52/public/app/features/variables/state/actions.ts#L86

Just give me a ping if you need any help getting this in place 🙏

aidanmountford committed:
Updated solution to retrieving $__user variables to pull directly from user record (thereby allowing future access to properties that
might not exist in the contextSvr).

Replicated this initialisation in the variables feature

Corrected typo's in documentation.

aidanmountford committed:
Repaired typecheck issues.

torkelo reviewed (commented):
Worried about the security implications of having a user variable. 

And this implementation is modifying the old template variable system (currently inactive), and adding api calls to fetch user which is not needed here.

aidanmountford commented:
@torkelo, specifically what security implications ?

A

On Wed, Apr 8, 2020 at 9:45 PM Torkel Ödegaard <notifications@github.com>
wrote:

> *@torkelo* commented on this pull request.
>
> Worried about the security implications of having a user variable.
>
> And this implementation is modifying the old template variable system
> (currently inactive), and adding api calls to fetch user which is not
> needed here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grafana/grafana/pull/23378#pullrequestreview-389896368>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACR6BDSQ42CEOZQUZ5FTJE3RLRPVBANCNFSM4MCSNW7Q>
> .
>


-- 

*Aidan Mountford*
Co-Founder & CEO
Fuuzio Group Pty Ltd
T1.509-514, 55 Plaza Parade
Maroochydore, QLD, 4557
Mobile: 0417-849-852

-- 

**This email message is for the sole use of the intended recipient(s) and 
may contain confidential and privileged information. Any unauthorized 
review, use, disclosure or distribution is prohibited. If you are not the 
intended recipient, please contact the sender by reply email and destroy 
all copies of the original message.**
**
**

**All electrical contracting 
activities are undertaken by Fuuzio Operations Pty Ltd.  ABN 38 627 150 547 
 (Registered Electrical Contractor **NSW:333365C,  VIC:28883, SA:294101).   
 **
**
**
**Fuuzio Operations Pty. Ltd. and Foreman-Sheean Electrical & 
Communications Pty. Ltd. are wholly owned subsidiaries of Fuuzio Group Pty. 
Ltd.**

mckn commented:
> Worried about the security implications of having a user variable.
> 
> And this implementation is modifying the old template variable system (currently inactive), and adding api calls to fetch user which is not needed here.

I share your worries but after talking to @hugohaggmark I realized that these values are already accessible via other APIs.

hugohaggmark commented on public/app/features/templating/variable_srv.ts:
```diff
@@ -65,12 +67,27 @@ export class VariableSrv {
             },
           },
         });
+        getBackendSrv()
+          .get('/api/users/' + contextSrv.user.id)
+          .then((user: User) => {
+            this.templateSrv.setGlobalVariable('__user', {
+              value: {
+                name: user.name,
+                login: user.login,
+                email: user.email,
+                id: contextSrv.user.id,
+                toString: function() {
+                  return this.id;
+                },
+              },
+            });
+          });
```
Do we really need to do this call? Isn't all this information already available in GrafanaBootData? ping @torkelo

aidanmountford replied:
@hugohaggmark 

I have changed so this is all pulled from contextSrv  (and added email/name to contextSrv).   

My original thinking here was to expand this to include the ability to pull info from a variety of API structures.   For now - being able to use the users login name/email in a query allows a dataset to be reduced to only those rows related to the individual user.   (In essence allowing a personalised dashboard)

The user information is available to the client post login both via an API, and in BootData.   One assumes that an administrator will not create users that have a name/email etc that are so structured that they permit SQL/code injection.

torkelo commented:
Accessible how? As template variables? There is a big difference between being accessible in plugin api vs template variable

aidanmountford committed:
Merge branch 'master' of https://github.com/grafana/grafana

aidanmountford committed:
Updated patch to pull entirely from contextSrv without API calls.

aidanmountford committed:
... And removed the redundant comments.

aidanmountford committed:
Updated documentation.

mckn commented:
> Accessible how? As template variables? There is a big difference between being accessible in plugin api vs template variable

You can fetch that information via a global variable on window. It isn't an API but when it comes to accessing this kind of information that might not matter.

aidanmountford commented:
Guys,

It is unclear where you would like me to take this, pls advise.   

At the moment this information is accessible within the client - It appears illogical to suggest that there is a security implication created from making data that is already available, available through a different vector.   

A

hugohaggmark commented:
So sorry for the bad responsiveness @aidanmountford , we're in the middle of closing Grafana 7.0 beta 2 and stable.

mckn commented:
> Guys,
> 
> It is unclear where you would like me to take this, pls advise.
> 
> At the moment this information is accessible within the client - It appears illogical to suggest that there is a security implication created from making data that is already available, available through a different vector.
> 
> A

Hello again,

Sorry for the delay. As @hugohaggmark said, we have been busy with the 7.0 release but let's see if we can get this one into master in the beginning of next week.

Would be super if you could sync with master until then? :pray:

aidanmountford committed:
Merge remote-tracking branch 'upstream/master'
(Post Grafana 7.0 release)
